### ISSUE TITLE
[feature request] same behavior in multi-gpu DataParallel vs single GPU

### ISSUE THREAD

**Issue by wenwei202** [pytorch/pytorch#5778]
In `DataParallel`, gradients from GPUs are **summed** to update parameters. This design results in the subtlety that training convergence depends on the number of GPUs you used to train, which makes experiments less duplicable and trackable when the number of gpus is different in hardware systems or when it is controlled by augment like `--gpus 0,1`. Is it better to **average** gradients from all GPUs like `torch.nn.parallel.DistributedDataParallel` does, so that the training convergence is independent on the number of gpus? Thanks!

**Comment by colesbury** [comment]
It's summed precisely because that makes the gradients independent from the number of GPUs. The gradients may depend on the batch size depending on the loss function you use.

See, for example:

```python
import torch
input = torch.randn(4, 128, 128).cuda()
linear = torch.nn.Linear(128, 128).cuda()
dp = torch.nn.DataParallel(linear, [0, 1])

linear(input).sum().backward()
print(linear.weight.grad)

linear.zero_grad()
dp(input).sum().backward()
print(linear.weight.grad)
```

If you use BatchNorm, your gradients will also depend on the number of GPUs because batch norm is computed per-GPU.

**Comment by wenwei202** [comment]
Thanks for quick reply. You are right, that's going to be equivalent if the loss function is a sum. However, CrossEntropyLoss averages gradients over the batch. If I want to make them equivalent, Can I do this so far:
```
model = torch.nn.DataParallel(model, args.gpus).cuda()
...
criterion = nn.CrossEntropyLoss().cuda()
...
for i, (input, target) in enumerate(train_loader):
        target = target.cuda(async=True)
        input_var = torch.autograd.Variable(input)
        target_var = torch.autograd.Variable(target)

        # compute output
        output = model(input_var)
        loss = criterion(output, target_var)

        # compute gradient and do SGD step
        optimizer.zero_grad()
        loss.backward()
        if not args.distributed:
            for p in model.parameters():
                p.grad.data.div_(len(args.gpus))
        optimizer.step()
```
Is it right?

**Comment by colesbury** [comment]
I may be missing something, but I don't see why you would divide by the number of GPUs there. The loss divides by the full batch size, not the number of GPUs.

**Comment by wenwei202** [comment]
Because [CrossEntropyLoss](http://pytorch.org/docs/master/nn.html#torch.nn.CrossEntropyLoss) by default has `size_average=True`, and the gradients of each GPU are already averaged  before synchronization/summation?

**Comment by colesbury** [comment]
Your criterion call is outside the data parallel. The size_average divides by the *full* batch size. There's still no dependence on the number of GPUs:

```python
import torch
import torch.nn as nn
from torch.autograd import Variable
input = Variable(torch.randn(8, 10).cuda())
target = Variable(torch.ones(8).long().cuda())
criterion = nn.CrossEntropyLoss().cuda()

linear = nn.Linear(10, 10).cuda()
dp = nn.DataParallel(linear, [0, 1])

criterion(linear(input), target).backward()
print(linear.weight.grad)

dp.zero_grad()
criterion(dp(input), target).backward()
print(linear.weight.grad)
```

**Comment by wenwei202** [comment]
Ooooooh, that's the problem I have always had. Thank you so much for it! Closing!
